ops/model.py

```diff
@@ -810,6 +901,291 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+        if id is not None:
+            id = self._canonicalize_id(id)
+        self._backend = backend
+        self._id = id
+        self._label = label
+        self._content = content
+
+    def __repr__(self):
+        fields = []  # type: List[str]
+        if self._id is not None:
+            fields.append('id={!r}'.format(self._id))
+        if self._label is not None:
+            fields.append('label={!r}'.format(self._label))
+        return '<Secret ' + ' '.join(fields) + '>'
+
+    @classmethod
+    def _canonicalize_id(cls, id: str) -> str:
+        """Return the canonical form of the given secret ID, with the 'secret:' prefix."""
+        id = id.strip()
+        if not id.startswith('secret:'):
+            id = 'secret:' + id  # add the prefix if not there already
+        return id
+
+    @classmethod
+    def _validate_content(cls, content: Optional[Dict[str, str]]):
+        """Ensure the given secret content is valid, or raise ValueError."""
+        if not isinstance(content, dict):
+            raise TypeError(f'Secret content must be a dict, not {type(content).__name__}')
+        if not content:
+            raise ValueError('Secret content must not be empty')
+
+        invalid_keys: List[str] = []
+        invalid_value_keys: List[str] = []
+        invalid_value_types: Set[str] = set()
+        for k, v in content.items():
+            if not cls._key_re.match(k):
+                invalid_keys.append(k)
+            if not isinstance(v, str):
+                invalid_value_keys.append(k)
+                invalid_value_types.add(type(v).__name__)
+
+        if invalid_keys:
+            raise ValueError(
+                f'Invalid secret keys: {invalid_keys}. '
+                f'Keys should be lowercase letters and digits, at least 3 characters long, '
+                f'start with a letter, and not start or end with a hyphen.')
+
+        if invalid_value_keys:
+            invalid_types = ' or '.join(sorted(invalid_value_types))
+            raise TypeError(f'Invalid secret values for keys: {invalid_value_keys}. '
+                            f'Values should be of type str, not {invalid_types}.')
+
+    @property
+    def id(self) -> Optional[str]:
+        """Unique identifier for this secret.
+
+        This will be None if you obtained the secret using
+        :meth:`Model.get_secret` with a label but no ID.
+        """
+        return self._id
+
+    @property
+    def label(self) -> Optional[str]:
+        """Label used to reference this secret locally.
+
+        This label is locally unique, that is, Juju will ensure that the
+        entity (the owner or consumer) only has one secret with this label at
+        once.
+
+        This will be None if you obtained the secret using
+        :meth:`Model.get_secret` with an ID but no label.
+        """
+        return self._label
+
+    def get_content(self, *, refresh: bool = False) -> Dict[str, str]:
```

Comment:
As a developer I'd rather be using `secret.content` and `secret.info` instead of calling `secret.get_content()` and `secret.get_info()`. And since refresh is discretionary to the user I believe it's fine to run something like

```
secret.refresh()
print(secret.content)
```

Reply:
Since the revision concept seems core the the idea of secrets, I wouldn't abstract it too much to the user.

Reply:
revision is only visible to the owner of a secret. consumers of a secret cannot see the 'revision' attribute, their only option is to stick with the current content or 'refresh' to new content, leaving the old behind. (they cannot, for example, ask for 'exactly revision 10 of the content). Hence why it is pretty hidden in the consumer objects.

Reply:
Thanks for the comments, folks. Yeah, I think there are enough problem with properties here, given that some properties are not settable, some are owner and some consumer, etc. So I'm going to stick with what we have. Not perfect, but clear about what you can set and when hook tool calls are made.